cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java

```diff
@@ -85,13 +87,19 @@ public void chunk(final ChannelBuffer request, HttpResponder responder) {
 
   @Override
   public void finished(HttpResponder responder) {
+    TransactionControl txCtrl = Transactions.getTransactionControl(
+      TransactionControl.IMPLICIT, HttpContentConsumer.class, delegate, "onFinish", HttpServiceResponder.class);
     try {
-      transactional.execute(new TxRunnable() {
-        @Override
-        public void run(DatasetContext context) throws Exception {
-          delegate.onFinish(BodyConsumerAdapter.this.responder);
-        }
-      });
+      if (TransactionControl.IMPLICIT == txCtrl) {
+        transactional.execute(new TxRunnable() {
+          @Override
+          public void run(DatasetContext context) throws Exception {
+            delegate.onFinish(BodyConsumerAdapter.this.responder);
+          }
+        });
+      } else {
+        delegate.onFinish(BodyConsumerAdapter.this.responder);
```

Comment:
Let's deal with this in a separate JIRA to revisit all calls to user methods, both with and without TX to make sure the switch is done correctly.